Issue 258 basexppauli class skeleton

dhruvbhq opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
Fix for issue #258 . Added skeleton class `BaseXPPauli` and methods `xp_pauli_rep.py`


### Details and comments
- Added skeleton class and methods in qiskit_qec/operators and qiskit_qec/utils
- Most function declarations from `BasePauli` and `pauli_rep.py` have been retained as of now. Functions are empty. There are placeholders for future documentation.
- Locally disabled lint warnings for newly added files
- Fixed a typo

dsvandet commented on qiskit_qec/operators/base_xp_pauli.py:
```diff
@@ -0,0 +1,460 @@
+# pylint: disable=no-member
+class BaseXPPauli(BaseOperator, AdjointMixin, MultiplyMixin):
+    r"""Symplectic representation of a list of N-qubit XP operators with phases using
+    numpy arrays for symplectic matrices and phase vectors.
```
... using numpy arrays for generalized symplectic matrices ...

dsvandet commented on qiskit_qec/operators/base_xp_pauli.py:
```diff
@@ -0,0 +1,460 @@
+        matrix: Union[np.ndarray, None] = None,
+        phase_exp: Union[None, np.ndarray, np.integer] = None,
+        order: str = "xz",
+    ) -> None:
```
A precision variable N is needed for XP Operators

dsvandet commented on qiskit_qec/operators/base_xp_pauli.py:
```diff
@@ -0,0 +1,460 @@
+        phase_exp: Union[None, np.ndarray, np.integer] = None,
+        order: str = "xz",
+    ) -> None:
+        """A BaseXPPauli object represents a list N-qubit Pauli operators with phases.
```
A BaseXPPauli object represents a list N-qubit **XP** Pauli operators

dsvandet commented on qiskit_qec/operators/base_xp_pauli.py:
```diff
@@ -0,0 +1,460 @@
+    ) -> None:
+        """A BaseXPPauli object represents a list N-qubit Pauli operators with phases.
+        Numpy arrays are used to represent the symplectic matrix represention of these
+        Paulis. The phases of the Paulis are stored encoded. The phases of the Pauli
```
Leave a TODO here as we need to work out the representations to be used for the XP operators. Also the

` (-i)^phase_exp Z^z X^x` 
 
 needs to be updated to XP operators etc.

dhruvbhq commented:
@dsvandet Thanks, I implemented the suggestions. Just one concern - `N` is used for the precision variable, at the same time, many docstrings say something like `...N-qubit Pauli operator...` etc. Could it be a potential source of confusion?

dsvandet commented:
Good point. Let's call it `precision` instead of `N`.

dsvandet commented:
In terms of the naming PauliXP  there are cases where  the XP operators no longer include the original Paulis. I view the XP operators  as 'generalized Paulis' in terms of not what the specifically are but in terms of their role and definition. From this point of view we do not need to always being including the original Paulis. In most cases the precision is going to be chosen such that the original Pauls are included. So let's leave this for the moment. We can always change later if a good reason shows itself.

dhruvbhq commented:
Thanks for your answer @dsvandet , it makes a lot of sense to me.
I've renamed `N` to `precision` and also changed the datatype `np.int8` to `np.int64` in `BaseXPPauli`, as discussed.

grace-harper-ibm reviewed (approved):
LGTM

dsvandet reviewed (approved):
We will want to revisit the use of np.integer. The reason is that np.integer can be very slow to check. Not important at this point so let's not change it now.
